### ISSUE TITLE
SyncProducer.SendMessages: handleSuccesses is attempting to send on nil expectation chan

### ISSUE THREAD

**Issue by timoha** [Shopify/sarama#1264]
##### Versions

Sarama Version: 1.12
Kafka Version: 2.0.0
Go Version: 1.11.4

##### Configuration
sarama:
```
    config := sarama.NewConfig()
    config.Version = sarama.V0_11_0_0
    config.Producer.Compression = sarama.CompressionSnappy
    config.Producer.Return.Successes = true
    config.Producer.Return.Errors = true
    config.Producer.RequiredAcks = sarama.WaitForAll
    config.Producer.Partitioner = sarama.NewManualPartitioner
```
broker:
```
port=9094
listeners=PLAINTEXT://0.0.0.0:9094
reserved.broker.max.id=1000000000
min.insync.replicas=2
default.replication.factor=3
unclean.leader.election.enable=false
num.replica.fetchers=4
replica.fetch.max.bytes=1048576
replica.fetch.wait.max.ms=500
replica.high.watermark.checkpoint.interval.ms=5000
replica.socket.timeout.ms=30000
replica.socket.receive.buffer.bytes=65536
replica.lag.time.max.ms=10000
replica.lag.max.messages=4000
controller.socket.timeout.ms=30000
controller.message.queue.size=10
num.partitions=8
message.max.bytes=1000000
auto.create.topics.enable=true
delete.topic.enable=true
offsets.topic.num.partitions=100
log.dir=/kafka-data/data
log.index.interval.bytes=4096
log.index.size.max.bytes=10485760
log.retention.hours=168
offsets.retention.minutes=10080
log.flush.interval.ms=10000
log.flush.interval.messages=20000
log.flush.scheduler.interval.ms=2000
log.roll.hours=168
log.retention.check.interval.ms=300000
log.segment.bytes=1073741824
zookeeper.connection.timeout.ms=6000
zookeeper.connect=zookeeper-0-zookeeper:2181,zookeeper-1-zookeeper:2181,zookeeper-2-zookeeper:2181/kafka
zookeeper.sync.time.ms=2000
num.io.threads=8
socket.request.max.bytes=104857600
socket.receive.buffer.bytes=1048576
socket.send.buffer.bytes=1048576
queued.max.requests=16
fetch.purgatory.purge.interval.requests=100
producer.purgatory.purge.interval.requests=100
```

##### Logs

```
I0124 04:23:10.676601       7 async_producer.go:855] producer/broker/8 state change to [retrying] on postdb_v1/7 because kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.
I0124 04:23:10.676651       7 async_producer.go:870] Retrying batch for postdb_v1-7 because of kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.
I0124 04:23:10.676673       7 async_producer.go:543] producer/leader/postdb_v1/7 state change to [retrying-1]
I0124 04:23:10.676719       7 async_producer.go:669] producer/broker/9 starting up
I0124 04:23:10.676768       7 async_producer.go:696] producer/broker/8 state change to [closed] on postdb_v1/7
I0124 04:23:10.676796       7 async_producer.go:553] producer/leader/postdb_v1/7 abandoning broker 8
I0124 04:23:10.676838       7 async_producer.go:748] producer/broker/8 shut down
I0124 04:23:10.677137       7 async_producer.go:904] producer/broker/9 state change to [closing] because EOF
I0124 04:23:10.677195       7 broker.go:191] Closed connection to broker 10.0.0.74:9094
I0124 04:23:10.776965       7 client.go:722] client/metadata fetching metadata for [postdb_v1] from broker 10.0.15.112:9094
I0124 04:23:10.777306       7 client.go:749] client/metadata got error from broker 1 while fetching metadata: EOF
I0124 04:23:10.777433       7 broker.go:191] Closed connection to broker 10.0.15.112:9094
I0124 04:23:10.777460       7 client.go:522] client/brokers deregistered broker #1 at 10.0.15.112:9094
I0124 04:23:10.777490       7 client.go:722] client/metadata fetching metadata for [postdb_v1] from broker 10.0.6.62:9094
I0124 04:23:10.778258       7 client.go:500] client/brokers registered new broker #1 at 10.0.15.112:9094
I0124 04:23:10.778340       7 async_producer.go:669] producer/broker/9 starting up
I0124 04:23:10.778368       7 async_producer.go:680] producer/broker/9 state change to [open] on postdb_v1/7
I0124 04:23:10.778385       7 async_producer.go:531] producer/leader/postdb_v1/7 selected broker 9
I0124 04:23:10.778470       7 async_producer.go:559] producer/leader/postdb_v1/7 state change to [flushing-1]
I0124 04:23:10.778486       7 async_producer.go:581] producer/leader/postdb_v1/7 state change to [normal]
I0124 04:23:10.778500       7 async_producer.go:543] producer/leader/postdb_v1/7 state change to [retrying-2]
I0124 04:23:10.778519       7 async_producer.go:553] producer/leader/postdb_v1/7 abandoning broker 9
I0124 04:23:10.778902       7 broker.go:146] Connected to broker at 10.0.0.74:9094 (registered as #9)
I0124 04:23:10.783014       7 async_producer.go:748] producer/broker/9 shut down
I0124 04:23:10.878666       7 client.go:722] client/metadata fetching metadata for [postdb_v1] from broker 10.0.2.19:9094
I0124 04:23:10.878879       7 client.go:749] client/metadata got error from broker 10 while fetching metadata: EOF
I0124 04:23:10.878964       7 broker.go:191] Closed connection to broker 10.0.2.19:9094
I0124 04:23:10.878980       7 client.go:522] client/brokers deregistered broker #10 at 10.0.2.19:9094
I0124 04:23:10.878992       7 client.go:722] client/metadata fetching metadata for [postdb_v1] from broker 10.0.0.74:9094
I0124 04:23:10.879673       7 client.go:500] client/brokers registered new broker #10 at 10.0.2.19:9094
I0124 04:23:10.879762       7 async_producer.go:669] producer/broker/9 starting up
I0124 04:23:10.879784       7 async_producer.go:680] producer/broker/9 state change to [open] on postdb_v1/7
I0124 04:23:10.879803       7 async_producer.go:531] producer/leader/postdb_v1/7 selected broker 9
```

##### Problem Description

We had some instability in our cluster which caused some of the brokers' zookeeper leases to expire (unfortunately don't have logs anymore, might be able to get some again). This lead to sarama.SyncProducer.SendMessages() to get stuck waiting on reading on `expectation` channel. Upon examining the stack trace, I've noticed that `handleSuccesses` is attempting to send on `ProducerMessage.expectation` channel that is nil:
```
goroutine 80 [chan send (nil chan), 892 minutes]:
vendor/github.com/Shopify/sarama.(*syncProducer).handleSuccesses(0xc000437a20)
        vendor/github.com/Shopify/sarama/sync_producer.go:133 +0x93
vendor/github.com/Shopify/sarama.(*syncProducer).handleSuccesses-fm()
        vendor/github.com/Shopify/sarama/sync_producer.go:76 +0x2a
vendor/github.com/Shopify/sarama.withRecover(0xc00041ced0)
        vendor/github.com/Shopify/sarama/utils.go:45 +0x43
created by vendor/github.com/Shopify/sarama.newSyncProducerFromAsyncProducer
        vendor/github.com/Shopify/sarama/sync_producer.go:76 +0xb8
```

After checking code of `async_producer.go`, it seems like there are couple places there `ProducerMessage` is instantiated without creating `expectation` channel (mostly for control messages `fin`, `syn`, `shutdown`). I'm wondering if there's a possibility of those messages leaking into `Successes()` channel and causing the deadlock?

**Comment by timoha** [comment]
Indeed just caught `ProducerMessage` with `fin` flag set and `expectation` channel being nil:
```
I0128 05:28:58.551960       7 async_producer.go:669] producer/broker/1 starting up
I0128 05:28:58.551993       7 async_producer.go:680] producer/broker/1 state change to [open] on postdb_v1/11
I0128 05:28:58.552015       7 async_producer.go:531] producer/leader/postdb_v1/11 selected broker 1
I0128 05:28:58.552090       7 async_producer.go:559] producer/leader/postdb_v1/11 state change to [flushing-1]
I0128 05:28:58.552106       7 async_producer.go:581] producer/leader/postdb_v1/11 state change to [normal]
I0128 05:28:58.552124       7 async_producer.go:543] producer/leader/postdb_v1/11 state change to [retrying-2]
I0128 05:28:58.552137       7 async_producer.go:553] producer/leader/postdb_v1/11 abandoning broker 1
I0128 05:28:58.552699       7 broker.go:146] Connected to broker at 10.0.15.112:9094 (registered as #1)
I0128 05:28:58.645466       7 async_producer.go:998] returnSuccesses: msg with flags: msg: {Topic:postdb_v1 Key:<nil> Value:<nil> Headers:[] Metadata:<nil> Offset:455280333 Partition:11 Timestamp:0001-01-01 00:00:00 +0000 UTC retries:1 flags:2 expectation:<nil> sequenceNumber:0}, flags: [fin]
I0128 05:28:58.645513       7 async_producer.go:1002] returnSuccesses: nil expectation chan! msg: {Topic:postdb_v1 Key:<nil> Value:<nil> Headers:[] Metadata:<nil> Offset:455280333 Partition:11 Timestamp:0001-01-01 00:00:00 +0000 UTC retries:1 flags:2 expectation:<nil> sequenceNumber:0}, flags: [fin]
I0128 05:28:58.645548       7 async_producer.go:748] producer/broker/1 shut down
```
This seems like a bug: if control messages are returned, they can falsely cause assumption of success. 
Would filtering all of messages that have `flags != 0` be a right fix or the problem is somewhere deeper and they shouldn't even be making it to this point?

**Comment by eapache** [comment]
> the problem is somewhere deeper and they shouldn't even be making it to this point

This. Seems like a weird state-machine bug. From this most recent occurrence can you provide a bit more log history for `producer/leader/postdb_v1/11` specifically. What were its state transitions prior to `[flushing-1]`?

**Comment by timoha** [comment]
Right before `[flushing-1]`:
```
I0128 04:16:29.790737       7 async_producer.go:531] producer/leader/postdb_v1/11 selected broker 12
I0128 04:16:29.790764       7 async_producer.go:680] producer/broker/12 state change to [open] on postdb_v1/11
I0128 04:16:29.790870       7 async_producer.go:559] producer/leader/postdb_v1/11 state change to [flushing-1]
I0128 04:16:29.790895       7 async_producer.go:581] producer/leader/postdb_v1/11 state change to [normal]
I0128 04:23:03.535456       7 client.go:724] client/metadata fetching metadata for all topics from broker kafka.default:9092
...
I0128 05:28:58.449450       7 client.go:722] client/metadata fetching metadata for [postdb_v1] from broker kafka.default:9092
I0128 05:28:58.450538       7 async_producer.go:855] producer/broker/12 state change to [retrying] on postdb_v1/11 because kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.
I0128 05:28:58.450594       7 async_producer.go:543] producer/leader/postdb_v1/11 state change to [retrying-1]
I0128 05:28:58.450610       7 async_producer.go:870] Retrying batch for postdb_v1-11 because of kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.
I0128 05:28:58.450665       7 async_producer.go:696] producer/broker/12 state change to [closed] on postdb_v1/11
I0128 05:28:58.450696       7 async_producer.go:553] producer/leader/postdb_v1/11 abandoning broker 12
I0128 05:28:58.468415       7 async_producer.go:904] producer/broker/1 state change to [closing] because EOF
I0128 05:28:58.477029       7 broker.go:191] Closed connection to broker 10.0.15.112:9094
I0128 05:28:58.550912       7 client.go:722] client/metadata fetching metadata for [postdb_v1] from broker kafka.default:9092
I0128 05:28:58.551960       7 async_producer.go:669] producer/broker/1 starting up
I0128 05:28:58.551993       7 async_producer.go:680] producer/broker/1 state change to [open] on postdb_v1/11
I0128 05:28:58.552015       7 async_producer.go:531] producer/leader/postdb_v1/11 selected broker 1
I0128 05:28:58.552090       7 async_producer.go:559] producer/leader/postdb_v1/11 state change to [flushing-1]
```

**Comment by timoha** [comment]
Tried to follow the logs and code. It looks like there's a race condition as broker 1 was closing and opening at the same time. 

`fin` message has `retries: 1`, which means `highWatermark` was 2 (https://github.com/Shopify/sarama/blob/master/async_producer.go#L550). Which corresponds to:
```
I0128 05:28:58.552124       7 async_producer.go:543] producer/leader/postdb_v1/11 state change to [retrying-2]
I0128 05:28:58.552137       7 async_producer.go:553] producer/leader/postdb_v1/11 abandoning broker 1
```
So this leaked `fin` message was due to broker 1. And it was sent to `bp.input` channel.

Then, I noticed that before we had:
```
I0128 05:28:58.468415       7 async_producer.go:904] producer/broker/1 state change to [closing] because EOF
```
which sets `closing` to non-nil. That's why we didn't discard `fin` here:
https://github.com/Shopify/sarama/blob/master/async_producer.go#L693

Can't really find a place where closing becomes non-nil. Maybe it shouldn't need to because the broker is meant to be garbage collected anyway, but it this case we ended up reopening it.
